cpp/src/experimental/graph.cu

```diff
@@ -0,0 +1,543 @@
+                                                  &number_of_local_edges_sum,
+                                                  1,
+                                                  raft::comms::op_t::SUM,
+                                                  default_stream);
+    CUGRAPH_EXPECTS(number_of_local_edges_sum == this->get_number_of_edges(),
+                    "Invalid API parameter: the sum of local edges doe counts not match with "
+                    "number_of_local_edges.");
+
+    detail::check_vertex_partition_offsets(partition.vertex_partition_offsets,
+                                           this->get_number_of_vertices());
+  }
+
+  // convert edge list (COO) to compressed sparse format (CSR or CSC)
+
+  adj_matrix_partition_offsets_.reserve(edgelists.size());
+  adj_matrix_partition_indices_.reserve(edgelists.size());
+  adj_matrix_partition_weights_.reserve(this->is_weighted() ? edgelists.size() : 0);
+  for (size_t i = 0; i < edgelists.size(); ++i) {
+    CUGRAPH_EXPECTS((is_weighted == false) || (edgelists[i].p_edge_weights != nullptr),
+                    "Invalid API parameter, edgelists[i].p_edge_weights shoud not be nullptr if "
+                    "is_weighted == true");
+
+    auto row_first = partition_.hypergraph_partitioned
+                       ? partition_.vertex_partition_offsets[comm_p_row_size * i + comm_p_row_rank]
+                       : partition_.vertex_partition_offsets[comm_p_col_size * comm_p_row_rank];
+    auto row_last =
+      partition_.hypergraph_partitioned
+        ? partition_.vertex_partition_offsets[comm_p_row_size * i + comm_p_row_rank + 1]
+        : partition_.vertex_partition_offsets[comm_p_col_size * (comm_p_row_rank + 1)];
+    auto col_first = partition_.vertex_partition_offsets[comm_p_row_size * comm_p_col_rank];
+    auto col_last  = partition_.vertex_partition_offsets[comm_p_row_size * (comm_p_col_rank + 1)];
+
+    rmm::device_uvector<edge_t> offsets(0, this->get_handle_ptr()->get_stream());
+    rmm::device_uvector<vertex_t> indices(0, this->get_handle_ptr()->get_stream());
+    rmm::device_uvector<weight_t> weights(0, this->get_handle_ptr()->get_stream());
+    std::tie(offsets, indices, weights) = edge_list_to_compressed_sparse<store_transposed>(
+      *(this->get_handle_ptr()), edgelists[i], row_first, row_last, col_first, col_last);
+    adj_matrix_partition_offsets_.push_back(std::move(offsets));
+    adj_matrix_partition_indices_.push_back(std::move(indices));
+    if (this->is_weighted()) { adj_matrix_partition_weights_.push_back(std::move(weights)); }
+  }
+
+  // update degree-based segment offsets (to be used for graph analytics kernel optimization)
+
+  if (sorted_by_global_degree_within_vertex_partition) {
+    auto degrees = detail::compute_major_degree(
+      *(this->get_handle_ptr()), adj_matrix_partition_offsets_, partition_);
+
+    // optinoal expensive checks (part 2/3)
+
+    if (do_expensive_check) {
```

Comment:
Perhaps rather than always check flags like this at runtime, a template policy `typename validity_checker_t = pass_through_check_t` could be added. It encapsulates all such (expensive) checks; and it defaults to a "pass through" policy that does nothing (i.e., all the corresponding member function checkers are "no-op"'s).

Reply:
This runtime check is very cheap (I mean runtime check of the flag, not the actual checks inside the if statement); so I think using the template policy here can be an unnecessary complication.

if (do_expensive_check) is executed only few times per graph object creation; this will add just some nanoseconds to the execution time. Graph object creation can take up to some seconds for large graphs. Some nanoseconds will be just noise.

Reply:
This isn't really about performance (although, calling these "expensive" suggest it could be). 

It's more about clean code. Unless the conditions under which such checks can change at runtime (in which case template policies cannot help...much) then using the policies would make the code cleaner. By "cleaner" I mean more functional-like and less if-then-else-y. Branching code makes it harder to read and maintain. 

Arguably, if different / more conditions pop-up in the future, it's easier for the maintainer of the code (likely other than the original creator) to just define corresponding policy functors and instantiate the class with them.

Which is easier to read / maintain, below:
```
//(1):
//
void foo(void* p, size_t sz) {
  if( sz == 32 )
    f1(*(static_cast<int32_t*>(p)));
  else if( sz == 64 )
    f2(*(static_cast<int64_t*>(p)));
  else
    f3(*(static_cast<long long*>(p)));
}
```
Or:

```
//(2):
//
template<typename T>
struct callable_t {
   void operator()(T t) {
     f3(t);
   }
};
template<> struct callable_t<int32_t> {
   void operator()(int32_t t) {
       f1(t);
    }
};
template<> struct callable_t<int64_t> {
   void operator()(int64_t t) {
       f2(t);
    }
};

template<typename T>
void foo(void* p)
{
   callable_t<T> f{};
   f(*(static_cast<T*>(p)));
}
```
I would argue (2) is easier to read and maintain (provided `T` can be dispatched at compile-time). What if you have to branch in (1) on `sz == 32` on the signed vs. unsigned case? You'd need to add branching logic inside the first branch, etc. While, with (2) you just supply a `uint32_t` specialization.

That being said, I leave this decision to you. If you feel strongly about the initial implementation then feel free to resolve this conversation.

Reply:
I think @seunghwak is trying to provide a runtime decision about whether to do these tests, not a compile time decision.  Looking at what's in the "expensive" tests, it's things like validating that the graph is, in fact, symmetric.  This (especially in MNMG mode) would be a very expensive test.  If the user wishes to assert/assume that the graph is in fact symmetric (they created it), they would want the option of being able to skip that validation test.

Reply:
Chuck's comment is pretty much to the point. Even for python users, this should be false by default, but if users need to debug, turning this expensive check flag on is the first thing they can try; this decision happens in runtime.